File: meeting/chat-service/src/domain/mappers/user.mapper.ts
import { UserDTO } from "../dto/user.dto";
import { User } from "../models/user.model";
import { IMapper } from "./mapper";
import { Document } from "mongoose";

export class UserMapper implements IMapper<UserDTO, User> {
    toDoc(dto: UserDTO): Partial<User> {
        return {
            _id: dto.id,
            avatarURL: dto.avatarURL,
            firstName: dto.firstName,
            lastName: dto.lastName,
        };
    }

    toDto(doc: User & Document): UserDTO {
        return {
            id: doc._id,
            avatarURL: doc.avatarURL,
            firstName: doc.firstName,
            lastName: doc.lastName,
        };
    }
}

File: account/user-service/src/main/java/org/example/userservice/mapper/UserMapper.java
package org.example.userservice.mapper;

import org.example.userservice.dto.UpdateUserDto;
import org.example.userservice.dto.UserDto;
import org.example.userservice.entity.Role;
import org.example.userservice.entity.User;
import org.example.userservice.service.contract.AvatarService;
import org.mapstruct.*;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Set;
import java.util.stream.Collectors;

@Mapper(componentModel = "spring")
public abstract class UserMapper {

    @Autowired
    protected AvatarService avatarService;

    @Mapping(target = "roles", source = "roles", qualifiedByName = "mapRolesToNames")
    @Mapping(target = "avatarUrl", source = "avatarKey", qualifiedByName = "mapAvatarKeyToUrl")
    public abstract UserDto toDto(User user);

    @Named("mapRolesToNames")
    protected Set<String> mapRolesToNames(Set<Role> roles) {
        if (roles == null) {
            return Set.of();
        }
        return roles.stream()
                .map(role -> role.getName().name())
                .collect(Collectors.toSet());
    }

    @Named("mapAvatarKeyToUrl")
    protected String mapAvatarKeyToUrl(String avatarKey) {
        if (avatarKey == null) return null;
        return avatarService.getAvatarUrl(avatarKey);
    }
}
